### ISSUE TITLE
Can we allow no-error as write event ? 

### ISSUE THREAD

**Issue by lambdai** [envoyproxy/envoy#18979]
The current Event::FileReadyType::Write is activate in 2 conditions

1. io handle is writable
2. io handle has error (bind error, or peer close)

The error type is delievered in the ConnectionImpl::immediate_error_event_ and [the immediate_error_event_ is initialized as Connected. ](https://github.com/envoyproxy/envoy/blob/main/source/common/network/connection_impl.h#L169)

Can we initialize this immediate event to support no-error value, e.g. absl::optional<ConnectionEvent> so that we can allow writable event be activated?

I have a simple commit in my local branch and it break nothing.

Motivation: I have a user space file event that can aggressively trigger write event.

Alternatively we can initialize the event after connect() or bind error occurs. 

CC @antoniovicente @yanavlasov @davinci26 for early comment.

**Comment by antoniovicente** [comment]
I'm having trouble understanding what you are proposing or why it is needed.  It is legal to call activate(Write).

The activate(Write) after setting immediate_error_event_ is to force a wakeup.  It could be a read activate instead.

**Comment by antoniovicente** [comment]
I think that the current implementation is the same as what you propose, Connected is used as a sentinel value that indicates NO_ERROR.

**Comment by lambdai** [comment]
Thank you for the feedback. 
The issue is that immediate_error_event_ doesn't support empty value.

If somehow activate(Write) is called, there are only two consequences:
Switch to connected from connecting, or switch to closed.  Here: https://github.com/envoyproxy/envoy/blob/main/source/common/network/connection_impl.cc#L673-L683
There is branch to keep the connection as "stay in connecting and wait for the next event"

**Comment by lambdai** [comment]
> I think that the current implementation is the same as what you propose, Connected is used as a sentinel value that indicates NO_ERROR.

Are you referring to [this branch](https://github.com/envoyproxy/envoy/blob/main/source/common/network/connection_impl.cc#L671-L679) ?
It's almost what I need except it switch the `connecting` state to false.

**Comment by antoniovicente** [comment]
You'll notice that ConnectionImpl::onWriteReady() does not depend on immediate_error_event_, so this member is not part of the problem.

Usual kernel socket semantics signal that a socket is connected by sending a Write event.  I recommend that your userspace socket send the first Write event once it considers itself connected, not earlier than that.

**Comment by lambdai** [comment]
> You'll notice that ConnectionImpl::onWriteReady() does not depend on immediate_error_event_, so this member is not part of the problem.

Right. onWriteReady is not issue.
I am referring to the branch onFileEvent here. https://github.com/envoyproxy/envoy/blob/main/source/common/network/connection_impl.cc#L556

It would be great to maintain the connection state that "ignoring all event" before connected. That's why I feel the alternative solution is to initiate the event only after listen or connect.

I can update user space event as well. But isn't it a better connection if it allow event notified? As if a read event triggered doesn't guarantee the following read() return data.

**Comment by lambdai** [comment]
I can create a sample commit see if it make sense to you, hold on...

**Comment by antoniovicente** [comment]
The scheduled event when immediate_error_event_ is set could be Read or Closed instead of Write since its purpose is just to trigger a wakeup after the callstack is unrolled.

A Write event typically signals that the socket transitioned to the connected state, for that reason it is not possible to ignore events before the socket is in the connected state.

I don't understand your proposal.

**Comment by lambdai** [comment]
@antoniovicente TL;DR is 

Set immediate_error_event_ when there is solid event.
1. after connect() is called if the connection is a client connection
2. or after a localclose/remoteclose.

See #18982

I also find an unexpected behavior that described in https://github.com/envoyproxy/envoy/issues/18981

**Comment by antoniovicente** [comment]
Commented on both.  Still don't understand your proposal.

**Comment by lambdai** [comment]
Both the the userspace event and os file event shares the same behavior: See #18993
The write event and read event is indeed activated before connect() is called.

**Comment by lambdai** [comment]
Given the fact above,  the client connection currently switch to connected state after a poll cycle even the connect() is not called.
The expected behavior is, a client connection should be in connecting state if there is no error.

As the next step, we should decide

1. Should a Envoy ConnectionImpl handle a activated  Read/Write event?  
  #18982 is following this path.

2. The connection owner must immediately calls connect() in the same poll cycle the connection is constructed.
  We don't have to do anything. We probably add comment or document.
3. other approach
  Like a connection factory?

WDTY @antoniovicente @davinci26

**Comment by lambdai** [comment]
#19860 creates a gap between tcp connection and quic connection.

A tcp connection must call connect() before the next event loop. See my test case https://github.com/envoyproxy/envoy/pull/18993

This inconsistent behavior is a tech debt:  either a connection or a connpool must be aware of the connection type.

Perhaps we should reconsider allow tcp connection call connect() at any poll cycle
